app/services/api/v3/parse_query_params_service.rb

```diff
@@ -30,50 +30,60 @@ module API
+        return json_parsed unless json_parsed.success?
 
-        parsed_params[:group_by] = group_by_from_params(params, skip_empty)
+        parsed = parsed_params(params)
 
-        error_result = with_service_error_on_json_parse_error do
-          parsed_params[:filters] = filters_from_params(params)
+        result = without_empty(parsed.merge(json_parsed.result), determine_allowed_empty(params))
 
-          parsed_params[:sort_by] = sort_by_from_params(params)
-
-          parsed_params[:timeline_labels] = timeline_labels_from_params(params)
-        end
-        return error_result if error_result
-
-        parsed_params[:columns] = columns_from_params(params)
-
-        parsed_params[:display_sums] = boolearize(params[:showSums])
-
-        parsed_params[:timeline_visible] = boolearize(params[:timelineVisible])
+        ServiceResult.new(success: true, result: result)
+      end
 
-        parsed_params[:timeline_zoom_level] = params[:timelineZoomLevel]
+      private
 
-        parsed_params[:highlighting_mode] = params[:highlightingMode]
+      def json_parsed_params(params)
+        parsed = {
+          filters: filters_from_params(params),
+          sort_by: sort_by_from_params(params),
+          timeline_labels: timeline_labels_from_params(params)
+        }
 
-        parsed_params[:highlighted_attributes] = highlighted_attributes_from_params(params)
+        ServiceResult.new success: true, result: parsed
+      rescue ::JSON::ParserError => error
+        result = ServiceResult.new success: false
+        result.errors.add(:base, error.message)
+        result
+      end
 
-        parsed_params[:display_representation] = params[:displayRepresentation]
+      def parsed_params(params)
+        {
+          group_by: group_by_from_params(params),
+          columns: columns_from_params(params),
+          display_sums: boolearize(params[:showSums]),
+          timeline_visible: boolearize(params[:timelineVisible]),
+          timeline_zoom_level: params[:timelineZoomLevel],
+          highlighting_mode: params[:highlightingMode],
+          highlighted_attributes: highlighted_attributes_from_params(params),
+          display_representation: params[:displayRepresentation],
+          show_hierarchies: boolearize(params[:showHierarchies])
+        }
+      end
 
-        parsed_params[:show_hierarchies] = boolearize(params[:showHierarchies])
+      def determine_allowed_empty(params)
+        allow_empty = params.keys
+        allow_empty << :group_by if group_by_empty?(params)
 
-        allow_empty = params.keys + skip_empty
-        ServiceResult.new(success: true, result: without_empty(parsed_params, allow_empty))
+        allow_empty
       end
 
-      def group_by_from_params(params, skip_empty)
+      # Group will be set to nil if parameter exists but set to empty string ('')
+      def group_by_from_params(params)
         return nil unless params_exist?(params, %i(group_by groupBy g))
```

Comment:
`group_by_keys` 1

Can we put `%i(group_by groupBy g)` into a method or constant and use that everywhere? I think I saw these 3 keys used in 3 different places. It may be unlikely that it will change but if it does we'll have to change it in those 3 places which we may forget.

E.g. I would change line 84 below (spot 2) from

```
attribute = params[:group_by] || params[:groupBy] || params[:g]
```
to
```
attribute = group_by_keys.find { |key| params[key] }
```